// @superfluid-finance/widget/dist/components/SuperfluidButton.js
import { jsx as _jsx } from "react/jsx-runtime";
import { Button, colors } from "@mui/material";
import SuperfluidLogo from "../SuperfluidLogo.js";
const SuperfluidButton = ({ widgetUrl, children, pulse = true, ...props }) => {
    return (_jsx(Button, { ...props, "data-testid": "go-to-widget-button", target: "_blank", variant: "contained", size: "large", href: widgetUrl, sx: {
            color: "white",
            textTransform: "none",
            ...(pulse
                ? {
                    animation: "pulse 3000ms ease-in-out infinite",
                }
                : {}),
            overflow: "hidden",
            whiteSpace: "nowrap",
            "@keyframes pulse": {
                "0%": {
                    transform: "scale(0.97.5)",
                    boxShadow: `0 0px 0px ${colors.green[500]}`,
                },
                "70%": {
                    transform: "scale(1)",
                    boxShadow: `0 0px 15px ${colors.green[500]}`,
                },
                "100%": {
                    transform: "scale(0.97.5)",
                    boxShadow: `0 0px 0px ${colors.green[300]}`,
                },
            },
        }, endIcon: _jsx(SuperfluidLogo, { fill: "#FFF" }), children: children ? children : "Pay with" }));
};
export default SuperfluidButton;
